app/services/teacher_training_public_api/sync_courses.rb

```diff
@@ -12,28 +12,97 @@ def perform(provider_id, recruitment_cycle_year)
+    end
+
+    def assign_course_attributes(course, course_from_api, recruitment_cycle_year)
+      course.name = course_from_api.name
+      course.level = course_from_api.level
+      course.study_mode = study_mode(course_from_api)
+      course.description = course_from_api.summary
+      course.start_date = course_from_api.start_date
+      course.course_length = course_from_api.course_length
+      course.recruitment_cycle_year = recruitment_cycle_year
+      course.exposed_in_find = course_from_api.findable
+      course.funding_type = course_from_api.funding_type
+      course.program_type = course_from_api.program_type
+      course.age_range = age_range_in_years(course_from_api)
+      course.withdrawn = course_from_api.state == 'withdrawn'
+      course.qualifications = course_from_api.qualifications
+    end
+
+    def study_mode(course_from_api)
+      course_from_api.study_mode == 'both' ? 'full_time_or_part_time' : course_from_api.study_mode
+    end
 
-      if course
-        course.update!(
-          program_type: course_from_api.program_type,
-          qualifications: course_from_api.qualifications,
-        )
+    def age_range_in_years(course_from_api)
+      if course_from_api.age_minimum.blank? || course_from_api.age_maximum.blank?
+        nil
+      else
+        "#{course_from_api.age_minimum} to #{course_from_api.age_maximum}"
       end
     end
+
+    def add_accredited_provider(course, accredited_body_code, recruitment_cycle_year)
+      if accredited_body_code.present? && course.provider.code != accredited_body_code
+
+        accredited_provider = ::Provider.find_by(code: accredited_body_code)
+
+        if accredited_provider.nil?
+          accredited_provider = create_new_accredited_provider(accredited_body_code, recruitment_cycle_year)
+        end
+
+        course.accredited_provider = accredited_provider
+        add_provider_relationship(course)
+
+      else
+        course.accredited_provider = nil
+      end
+    end
+
+    def add_provider_relationship(course)
+      ProviderRelationshipPermissions.find_or_create_by!(
+        training_provider: provider,
+        ratifying_provider: course.accredited_provider,
+      )
+    end
+
+    def create_new_accredited_provider(accredited_body_code, recruitment_cycle_year)
+      new_provider = TeacherTrainingPublicAPI::Provider
+                         .where(year: recruitment_cycle_year)
+                         .find(accredited_body_code).first
+
+      accredited_provider = ::Provider.new(code: accredited_body_code)
+      accredited_provider.code = new_provider.code
+      accredited_provider.name = new_provider.name
+      accredited_provider.region_code = new_provider.region_code&.strip
+      accredited_provider.postcode = new_provider.postcode
+      accredited_provider.provider_type = new_provider.provider_type
+      accredited_provider.latitude = new_provider.try(:latitude)
+      accredited_provider.longitude = new_provider.try(:longitude)
```

Comment:
Do we need to guard here withe the`try` method? Genuine question. I would have thought it would just return nil if one isn't present.

Reply:
Yes, I think we can. Based on the API docs it looks like they should be provided